Android/app/src/main/java/app/intra/IntroDialog.java

```diff
@@ -0,0 +1,187 @@
+    }
+    backButton.setOnClickListener(new View.OnClickListener() {
+      @Override
+      public void onClick(View view) {
+        int currentItem = pager.getCurrentItem();
+        if (currentItem > 0) {
+          pager.setCurrentItem(currentItem - 1);
+        }
+      }
+    });
+
+    Button nextButton = welcomeView.findViewById(R.id.intro_next);
+    Drawable endChevron = getResources().getDrawable(R.drawable.ic_chevron_end);
+    if (rtl) {
+      nextButton.setCompoundDrawablesWithIntrinsicBounds(endChevron, null, null, null);
+    } else {
+      nextButton.setCompoundDrawablesWithIntrinsicBounds(null, null, endChevron, null);
+    }
+    nextButton.setOnClickListener(new View.OnClickListener() {
+      @Override
+      public void onClick(View view) {
+        int currentItem = pager.getCurrentItem();
+        if (currentItem < NUM_PAGES - 1) {
+          pager.setCurrentItem(currentItem + 1);
+        }
+      }
+    });
+
+    Button acceptButton = welcomeView.findViewById(R.id.intro_accept);
+    acceptButton.setOnClickListener(new View.OnClickListener() {
+      @Override
+      public void onClick(View view) {
+        PersistentState.setWelcomeApproved(getContext(), true);
+        DnsVpnController.getInstance().start(getContext());
+        dismiss();
+      }
+    });
+
+    pager.addOnPageChangeListener(new TabListener(backButton, nextButton, acceptButton));
+
+    shown = true;
+    return welcomeView;
+  }
+
+  public static class Page0 extends Fragment {
+    @Override
+    public View onCreateView(LayoutInflater inflater, ViewGroup container,
+                             Bundle savedInstanceState) {
+      return inflater.inflate(R.layout.intro0, container, false);
+
+    }
+  }
+
+  public static class Page1 extends Fragment {
+    @Override
+    public View onCreateView(LayoutInflater inflater, ViewGroup container,
+                             Bundle savedInstanceState) {
+      return inflater.inflate(R.layout.intro1, container, false);
+    }
+  }
+
+  public static class Page2 extends Fragment {
+    @Override
+    public View onCreateView(LayoutInflater inflater, ViewGroup container,
+                             Bundle savedInstanceState) {
+      View view = inflater.inflate(R.layout.intro2, container, false);
+      TextView details = view.findViewById(R.id.intro_details_body);
+      details.setMovementMethod(LinkMovementMethod.getInstance());
+      return view;
+    }
+  }
+
+  private class Adapter extends FragmentPagerAdapter {
+    private final Fragment[] pages;
+    Adapter(FragmentManager fm) {
+      super(fm);
+      pages = new Fragment[NUM_PAGES];
+      pages[0] = new Page0();
+      pages[1] = new Page1();
+      pages[2] = new Page2();
+    }
+
+    @Override
+    public int getCount() {
+      return pages.length;
+    }
+
+    @Override
+    public Fragment getItem(int position) {
+      return pages[position];
```

Comment:
Maybe check that the index is in bounds?

Reply:
Done.